### PR TITLE
add UT for bypass

### LINKED ISSUE #371: Add UT for Bypass
**What would you like to be added**:
UT for bypass feature
**Why is this needed**:
There is currently no ut to ensure the quality of the bypass feature, and new use cases need to be added.

### PR DESCRIPTION
**Pull request by weli-l**
**What type of PR is this?**

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #371 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: pkg/controller/bypass/bypass_test.go [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+			Labels: map[string]string{
+				"kmesh.net/bypass": "enabled",
+			},
+		},
+		Spec: corev1.PodSpec{
+			NodeName: "test-node",
+		},
+	}
+
+	patches2.ApplyFunc(isKmeshManaged, func(pod *corev1.Pod) bool {
```
Comment (hzxuzhonghu) [inline pkg/controller/bypass/bypass_test.go:49]: nit we donot need to replace `isKmeshManaged`, which is simple checking pod annotations

File: pkg/controller/bypass/bypass_test.go [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+			wantErr: false,
+		},
+	}
+
+	go StartByPassController(client)
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			tt.beforeFunc()
+			time.Sleep(1 * time.Second)
```
Comment (hzxuzhonghu) [inline pkg/controller/bypass/bypass_test.go:120]: donot use sleep, it is very time costy. use async notify instead

File: pkg/controller/bypass/bypass_test.go [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+						},
+					},
+				},
+			},
+			want: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got, _ := checkSidecar(tt.args.client, tt.args.pod); got != tt.want {
```
Comment (hzxuzhonghu) [inline pkg/controller/bypass/bypass_test.go:208]: what does `client.CoreV1().Namespaces().Get` return, when you did not create a namespace

File: pkg/controller/bypass/bypass_test.go [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+			wantErr: false,
+		},
+	}
+
+	go StartByPassController(client)
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			tt.beforeFunc()
+			time.Sleep(1 * time.Second)
```
Comment (hzxuzhonghu) [inline pkg/controller/bypass/bypass_test.go:283]: ditto, prevent using sleep

File: pkg/controller/bypass/bypass_test.go [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+			},
+			afterFunc: func() {
+				assert.Equal(t, 1, addFuncCallCount, "AddFunc should not been called")
+				assert.Equal(t, 0, delFuncCallCount, "DelFunc should have been called once")
+			},
+			wantErr: false,
+		},
+	}
+
+	go StartByPassController(client)
```
Comment (LiZhenCheng9527) [inline pkg/controller/bypass/bypass_test.go:115]: Why is there a goroutine on this side. Is it possible that the use a goroutine on this side results in a 1s interval between tests?

File: pkg/controller/bypass/bypass_test.go [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+						},
+					},
+				},
+			},
+			want: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := isKmeshManaged(tt.args.pod); got != tt.want {
```
Comment (LiZhenCheng9527) [inline pkg/controller/bypass/bypass_test.go:168]:
```suggestion
			assert.Equal(t, isKmeshManaged(tt.args.pod), tt.want)
```

File: pkg/controller/bypass/bypass_test.go [hunk 7]
```diff
@@ -0,0 +1,287 @@
+package bypass
```
Comment (LiZhenCheng9527) [inline pkg/controller/bypass/bypass_test.go:1]: copyright

File: pkg/controller/bypass/bypass_test.go [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+	})
+
+	tests := []struct {
+		name       string
+		beforeFunc func()
+		afterFunc  func()
+		wantErr    bool
+	}{
+		{
+			name: "test1: pod with bypass label triggers enable bypass kmesh",
```
Comment (YaoZengzeng) [inline pkg/controller/bypass/bypass_test.go:70]: There is no need to add a prefix like "test1"

File: pkg/controller/bypass/bypass_test.go [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+	type args struct {
+		pod *corev1.Pod
+	}
+	tests := []struct {
+		name string
+		args args
+		want bool
+	}{
+		{
+			name: "test1: annotation with kmesh.net/redirection=enabled, should return true",
```
Comment (YaoZengzeng) [inline pkg/controller/bypass/bypass_test.go:138]: ditto

File: pkg/controller/bypass/bypass_test.go [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+		client kubernetes.Interface
+		pod    *corev1.Pod
+	}
+	tests := []struct {
+		name string
+		args args
+		want bool
+	}{
+		{
+			name: "test1: pod withsidecar inject annotation, should return true",
```
Comment (YaoZengzeng) [inline pkg/controller/bypass/bypass_test.go:188]:
```suggestion
			name: "test1: pod with sidecar inject annotation, should return true",
```

File: pkg/controller/bypass/bypass_test.go [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+	}
+
+	tests := []struct {
+		name       string
+		beforeFunc func()
+		afterFunc  func()
+		wantErr    bool
+	}{
+		{
+			name: "test1: pod with sidecar annotation triggers addIptables",
```
Comment (YaoZengzeng) [inline pkg/controller/bypass/bypass_test.go:255]: ditto

File: pkg/controller/bypass/bypass_test.go [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+				assert.Equal(t, 2, count, "utils.Excute should been called twice")
+			},
+			wantErr: false,
+		},
+		{
+			name: "test2: pod deletion triggers delIptables",
+			beforeFunc: func() {
+				count = 0
+				err := client.CoreV1().Pods("default").Delete(context.TODO(), "test-pod-with-label", metav1.DeleteOptions{})
+				require.NoError(t, err, "failed to create test pod")
```
Comment (YaoZengzeng) [inline pkg/controller/bypass/bypass_test.go:271]:
```suggestion
				require.NoError(t, err, "failed to delete test pod")
```

File: pkg/controller/bypass/bypass_test.go [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+				podWithLabel.DeletionTimestamp = &metav1.Time{Time: time.Now()}
+				_, err = client.CoreV1().Pods("default").Update(context.TODO(), podWithLabel, metav1.UpdateOptions{})
+				require.NoError(t, err, "failed to update test pod with deletion timestamp")
+
+				err = client.CoreV1().Pods("default").Delete(context.TODO(), "test-pod-with-label", metav1.DeleteOptions{})
+				require.NoError(t, err, "failed to delete test pod")
+			},
+			afterFunc: func() {
+				assert.Equal(t, 1, addFuncCallCount, "AddFunc should not been called")
+				assert.Equal(t, 0, delFuncCallCount, "DelFunc should have been called once")
```
Comment (YaoZengzeng) [inline pkg/controller/bypass/bypass_test.go:110]: The descriptions are right?

File: pkg/controller/bypass/bypass_test.go (hunk 7 above)
Comment (hzxuzhonghu) [reply pkg/controller/bypass/bypass_test.go:1]: Please add copyright

File: pkg/controller/bypass/bypass_test.go [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+
+	"github.com/agiledragon/gomonkey/v2"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/fake"
+)
+
+func waitAndCheckByPassAction(t *testing.T, wg *sync.WaitGroup, done chan struct{}, mu *sync.Mutex, enabled *bool, disabled *bool, enableExpected bool, disableExpected bool) {
```
Comment (hzxuzhonghu) [inline pkg/controller/bypass/bypass_test.go:18]: this function is now not needed, you can assert values after wg.Wait()

**Review by hzxuzhonghu — APPROVED** [review]
/lgtm

**Pull request merged**